graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -268,47 +304,128 @@ def enter(self, node, key, parent, path, ancestors):
+                # suppress the type
+                return REMOVE
             else:  # Make copy of node with the changed name, return the copy
                 field_node_with_new_name = get_copy_of_node_with_new_name(node, new_field_name)
                 return field_node_with_new_name
+
+
+class CascadingSuppressionCheckVisitor(Visitor):
+    def __init__(self, renamings: Dict[str, Optional[str]], query_type: str):
+        """Create a visitor to check that suppression does not cause an illegal state.
+
+        Args:
+            renamings: Dict[str, Optional[str]], from original field name to renamed field name. Any
+                       name not in the dict will be unchanged
+            query_type: str, name of the query type (e.g. RootSchemaQuery)
+        """
+        # Note that as field names and type names have been confirmed to match up, any renamed
+        # query type field already has a corresponding renamed type. If no errors, due to either
+        # invalid names or name conflicts, were raised when renaming type, no errors will occur when
+        # renaming query type fields.
+        self.in_query_type = False
+        self.renamings = renamings
+        self.query_type = query_type
+
+    def enter_object_type_definition(self, node: ObjectTypeDefinitionNode, *args):
+        """If the node's name matches the query type, record that we entered the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = True
+
+    def leave_object_type_definition(
+        self, node: ObjectTypeDefinitionNode, key, parent, path, ancestors
+    ):
+        """If the node's name matches the query type, record that we left the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = False
+
+    def enter_field_definition(self, node: FieldDefinitionNode, *args):
+        """If not at query type, check that no field depends on a type that was suppressed."""
+        if self.in_query_type:
+            return None
+        # at a field of a type that is not the query type
+        field_name = node.name.value
+        node_type = get_ast_with_non_null_and_list_stripped(node.type)
+        if node_type == REMOVE:
+            # then this field depends on a type that was suppressed, which is illegal
+            type_name = args[-1][-1].name.value
+            raise CascadingSuppressionError(
+                f"Type renamings {self.renamings} attempted to suppress a type, but type "
+                f"{type_name}'s field {field_name} still depends on that type. Suppressing "
+                f"individual fields hasn't been implemented yet, but when it is, you can fix "
+                f"this error by suppressing the field as well. Note that adding suppressions "
+                f"may lead to other types, fields, unions, etc. requiring suppression so you "
+                f"may need to iterate on this before getting a legal schema."
+            )
```

Comment:
Perhaps we can add a separate helper (separate PR, though perhaps build it and merge it first?) that given a list of types we'd _like to_ suppress, produces the list of fields etc. that must be suppressed as well for the type suppression to be legal. That enables a few nice things:
- The workflow becomes more pleasant: "iterate until you stop getting errors" becomes "use this helper, review its output, and apply it if it looks good".
- The invariant we check for here simply becomes "the helper finds no fields that need suppression", enabling code reuse.

Reply:
That's a really good point-- now that you mention it, we can do something similar with unions as well.

i.e. Instead of raising an error here, we could add the information about the field to a data structure belonging to this visitor object. Then, after `visit` returns, we could perform a single pass through this data structure and raise errors as needed for any fields/ unions, showing everything in a single pass _before_ `rename` calls `_rename_types` at all. 

I think this actually shouldn't be all that much work since most of it is already done with this visitor class so I can probably get it in this PR.